x/mongo/driver/topology/pool.go

```diff
@@ -825,12 +855,60 @@ func (p *pool) checkInNoEvent(conn *connection) error {
 	return nil
 }
 
+// clear calls clearImpl internally with a false interruptAllConnections value.
+func (p *pool) clear(err error, serviceID *primitive.ObjectID) {
+	p.clearImpl(err, serviceID, false)
+}
+
+// clearAll does same as the "clear" method but interrupts all connections.
+func (p *pool) clearAll(err error, serviceID *primitive.ObjectID) {
+	p.clearImpl(err, serviceID, true)
+}
+
+// interruptConnections interrupts the input connections.
+func (p *pool) interruptConnections(conns []*connection) {
+	for _, conn := range conns {
+		_ = conn.closeWithErr(poolClearedError{
+			err:     fmt.Errorf("interrupted"),
+			address: p.address,
+		})
+		_ = p.checkInWithCallback(conn, func() (reason, bool) {
```

Comment:
Checking in the connection to close it can be dangerous. If there is ever a case where the logic to check for perished connections has a bug, we may check-in an in-use connection, which could lead to data corruption. We should close the connections directly here and not rely on check-in to close them.